### PR TITLE
SCA hardening for modular field inversion in `EC_GROUP`s

### PR DESCRIPTION
**Pull request by romen**
# SCA hardening for modular field inversion in `EC_GROUP`s

This PR adds a dedicated function in `EC_METHOD` to access a modular field inversion implementation suitable for the specifics of the implemented curve, featuring SCA countermeasures.

The new pointer is defined as:
```c
int (*field_inv)(const EC_GROUP *, BIGNUM *r, const BIGNUM *a, BN_CTX *)
```
and computes the multiplicative inverse of `a` in the underlying field, storing the result in `r`.

Three implementations are included, each including specific SCA countermeasures:
  - `ec_GFp_simple_field_inv()`, featuring SCA hardening through blinding.
  - `ec_GFp_mont_field_inv()`, featuring SCA hardening through *Fermat's Little Theorem* (FLT) inversion.
  - `ec_GF2m_simple_field_inv()`, that uses `BN_GF2m_mod_inv()` which already features SCA hardening through blinding.

This PR also adds a new error reason code (i.e., `EC_R_CANNOT_INVERT`) to improve consistency between the three implementations as all of them could fail for the same reason but through different code paths resulting in inconsistent error stack states.

## Security analysis

Implementing this countermeasures also helps addressing a leakage affecting conversions from projective to affine coordinates in `EC_POINT_get_affine_coordinates()` which did not consider the `BN_FLG_CONSTTIME` flag, hitting a leaky code path for inversion.

Public disclosure of this minor defect has been responsibly coordinated with the OpenSSL security team.

## Backports

- This should go to `master` and `1.1.1-stable`.
- Ideally this could be backported to `1.1.0-stable` to minimize discrepancies between it and the LTS branch and minimize potential future security fixes, but technically I am not sure if this qualifies as security fix.
- Backport to 1.0.2 is not recommended, as it would require modifying non-opaque `struct`s.

## Checklist

- [x] documentation is added
- [x] tests are added

### CODE REVIEW

**Comment by mattcaswell** [comment]
I personally think this should go to 1.1.0 as well - although I suspect it may not cherry-pick cleanly due to the tests. Anyway I'll add the label for now.

File: crypto/ec/ec2_smpl.c [hunk 1]
```diff
@@ -889,6 +889,20 @@ int ec_GF2m_simple_points_mul(const EC_GROUP *group, EC_POINT *r,
...
 
+/*-
+ * Computes the multiplicative inverse of a in GF(2^m), storing the result in r.
+ * If a is zero (or equivalent), you'll get a EC_R_CANNOT_INVERT error.
+ * SCA hardening is with blinding: BN_GF2m_mod_inv does that.
+ */
+static int ec_GF2m_simple_field_inv(const EC_GROUP *group, BIGNUM *r,
+                                    const BIGNUM *a, BN_CTX *ctx)
+{
+    int ret;
```
Comment (mattcaswell) [inline crypto/ec/ec2_smpl.c]: Blank line after declarations

File: crypto/ec/ec_lcl.h [hunk 2]
```diff
@@ -154,6 +154,10 @@ struct ec_method_st {
     int (*field_sqr) (const EC_GROUP *, BIGNUM *r, const BIGNUM *a, BN_CTX *);
     int (*field_div) (const EC_GROUP *, BIGNUM *r, const BIGNUM *a,
                       const BIGNUM *b, BN_CTX *);
+    /* computes the multipicative inverse of a in the field, storing the result in r.
```
Comment (mattcaswell) [inline crypto/ec/ec_lcl.h]: Multi-line comment style. First line should be "/*" on its own.

File: crypto/ec/ecp_mont.c [hunk 3]
```diff
@@ -206,6 +207,54 @@ int ec_GFp_mont_field_sqr(const EC_GROUP *group, BIGNUM *r, const BIGNUM *a,
...
+        goto err;
+
+    /* Inverse in constant time with Fermats Little Theorem */
+    if (!BN_set_word(e, 2))
+        goto err;
+    if (!BN_sub(e, group->field, e))
+        goto err;
+    /*-
+     * Exponent e is public.
+     * No need for scatter-gather or BN_FLG_CONSTTIME.
```
Comment (mattcaswell) [inline crypto/ec/ecp_mont.c]: Not sure I understand the scatter-gather comment. Can you explain?

**Comment by mattcaswell** [comment]
Travis red cross is relevant. 

````
test/ec_internal_test.c:48:84: error: invalid operands to binary expression ('void' and 'int')
        || !test_true("test/ec_internal_test.c", 48, "BN_zero(a)", (BN_zero_ex(a)) != 0)
                                                                   ~~~~~~~~~~~~~~~ ^  ~
1 error generated.
make[1]: *** [test/ec_internal_test-bin-ec_internal_test.o] Error 1
make[1]: Leaving directory `/home/travis/build/openssl/openssl'
make: *** [build_programs] Error 2
````

That's quite interesting. `BN_zero` is a macro defined like this:
````
# if OPENSSL_API_0_9_8
#  define BN_zero(a)      BN_zero_ex(a)
# else
#  define BN_zero(a)      (BN_set_word((a),0))
# endif
````
So, depending on you build BN_zero() is either a void return or an int return!!!! That's not great. :-(

File: crypto/ec/ec_lcl.h (hunk 2 above)
Comment (romen) [reply crypto/ec/ec_lcl.h]: fixed! thanks!

File: crypto/ec/ec2_smpl.c (hunk 1 above)
Comment (romen) [reply crypto/ec/ec2_smpl.c]: fixed!

**Comment by romen** [comment]
> So, depending on you build BN_zero() is either a void return or an int return!!!! That's not great. :-(

Yep! I cringed when I noticed `BN_zero()` and `BN_one()` have different signatures!

Sidenote: is 3.0.0 a good candidate to make `BN_zero()` "normal"?

**Comment by mattcaswell** [comment]
> Sidenote: is 3.0.0 a good candidate to make `BN_zero()` "normal"?

TBH the BN_zero thing is a mess and I'm not sure what to do about it. AFAICT the default is you get:

````
#  define BN_zero(a)      (BN_set_word((a),0))
````

But if you build with no-deprecated you get a different definition for it with a different (void) return value:

````
#  define BN_zero(a)      BN_zero_ex(a)
````

I'm tempted to get rid of the no-deprecated form and revert back to the default (deprecated) form in all cases. At least that way you get a signature that is consistent with BN_one() and won't cause unneccessary breakage when we (at some point in the future) start removing deprecated stuff.

(BTW: Do you have a response to my scatter-gather comment above?)

File: crypto/ec/ecp_mont.c (hunk 3 above)
Comment (romen) [reply crypto/ec/ecp_mont.c]: It's @bbbrumley comment, but I would say that he meant that when inverting through FLT, we do `a^(p-2)`: the exponentiation could leak the exponent, which is public anyway, so there is no need to use `BN_FLG_CONSTTIME` to do a const-time exponentiation nor to somehow blind the scalar `a`.

**Comment by romen** [comment]
Travis red cross now is unrelated: there was some fetch issue on the osx builds.
Can we kick them to run again somehow?

> Do you have a view on the likely performance implications of this?

I will conjure some numbers, but I need more time to convince `speed` to hit all the relevant code paths.

File: crypto/ec/ecp_mont.c (hunk 3 above)
Comment (bbbrumley) [reply crypto/ec/ecp_mont.c]: @mattcaswell Read the comments in #6116

> To that end, [e913d11](https://github.com/openssl/openssl/commit/e913d11f444e0b46ec1ebbf3340813693f4d869d) was not really correct

**Comment by bbbrumley** [comment]
> Do you have a view on the likely performance implications of this?

Should be close to zero. See Sec. 7.2 [here](https://eprint.iacr.org/2018/651). It's in the ECC field not the group order, but similar sizes.

File: crypto/ec/ecp_mont.c (hunk 3 above)
Comment (bbbrumley) [reply crypto/ec/ecp_mont.c]: What's more, to have a chuckle at #8253. Performance wise, no one will notice if the flag is set or not. At some point #6116 was randomly inheriting the flag anyway. I'm not sure about the current state.

Comment (romen) [reply crypto/ec/ecp_mont.c]: > @mattcaswell Read the comments in #6116
> 
> > To that end, [e913d11](https://github.com/openssl/openssl/commit/e913d11f444e0b46ec1ebbf3340813693f4d869d) was not really correct

I had a hard time finding the thread in the comments of #6116. This should be the conversation @bbbrumley is pointing at: https://github.com/openssl/openssl/pull/6116#discussion_r186595991

Comment (davidben) [reply crypto/ec/ecp_mont.c]: Scatter-gather is probably not the best descriptor of `BN_mod_exp_mont` vs `BN_mod_exp_mont_consttime`. It's whether you use fixed windows with secret window values, or variable windows with public window values. Scatter-gather is a particular implementation strategy for the table lookup given secret window values. (And, depending on how it's done, [can be questionable](https://eprint.iacr.org/2016/224.pdf). Best to exhibit the same memory access pattern even within a cacheline.)

**Comment by romen** [comment]
> > Do you have a view on the likely performance implications of this?
> 
> Should be close to zero. See Sec. 7.2 [here](https://eprint.iacr.org/2018/651). It's in the ECC field not the group order, but similar sizes.

I can confirm with fresh data!

# Benchmark

- `nistp256`, using the `nistz256` implementation, is unaffected as it never hits any `ec_GF*_field_inv()`
- `nistp521`, representative of the `ec_nistp_64_gcc_128` implementation family, is unaffected as it never hits any `ec_GF*_field_inv()`
- `nistb571` uses the `EC_GF2m_simple_method`, hits the `ec_GF2m_simple_field_inv()` method, which is just a wrapper around the previous implementation which was hardened in #6070 (0dae8bafceabc8966383aa1f11ee8622f7dbde2f), so no real change in performance
- `brainpoolP512r1` uses the `EC_GFp_mont_method` and hits the `ec_GFp_mont_field_inv()` method with the FLT hardening: the comparison shows that the performance hit is insignificant
- hitting `ec_GFp_simple_field_inv()` from a named curve within the library is difficult: one way is using `nistp{192,224,256,384,521}` in builds with `no-asm`, which end up using the `EC_GFp_nist_method` which "inherits" the `ec_GFp_simple_field_inv()` function.
   Even in this case the performance backlash is minimal.

## Before this PR (fa1f03061037cbdac5369849a885c1191a2550d9)

```
> openssl speed ecdhb571 ecdhbrp512r1 ecdhp521 ecdhp256
version: 3.0.0-dev
built on: built on: Fri Feb 15 19:15:35 2019 UTC
options:bn(64,64) rc4(16x,int) des(int) aes(partial) idea(int) blowfish(ptr) 
compiler: ccache gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -g -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DNDEBUG
                              op      op/s
 256 bits ecdh (nistp256)   0.0001s  19587.7
 521 bits ecdh (nistp521)   0.0003s   2965.6
 571 bits ecdh (nistb571)   0.0019s    530.0
 512 bits ecdh (brainpoolP512r1)   0.0014s    715.8

> openssl speed ecdhp521 # no-asm build
version: 3.0.0-dev
built on: built on: Fri Feb 15 19:58:37 2019 UTC
options:bn(64,64) rc4(int) des(int) aes(partial) idea(int) blowfish(ptr) 
compiler: ccache gcc -fPIC -pthread -m64 -Wall -O3 -g -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DNDEBUG
                              op      op/s
 521 bits ecdh (nistp521)   0.0038s    262.2
```

## After this PR (ca65f5b373c5ed29b132fa1aa9a0ebbc1e874807)

```
> openssl speed ecdhb571 ecdhbrp512r1 ecdhp521 ecdhp256
version: 3.0.0-dev
built on: built on: Fri Feb 15 19:17:59 2019 UTC
options:bn(64,64) rc4(16x,int) des(int) aes(partial) idea(int) blowfish(ptr) 
compiler: ccache gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -g -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DNDEBUG
                              op      op/s
 256 bits ecdh (nistp256)   0.0001s  19684.2
 521 bits ecdh (nistp521)   0.0003s   2961.1
 571 bits ecdh (nistb571)   0.0019s    530.9
 512 bits ecdh (brainpoolP512r1)   0.0014s    716.1

> openssl speed ecdhp521 # no-asm build
version: 3.0.0-dev
built on: built on: Fri Feb 15 19:43:40 2019 UTC
options:bn(64,64) rc4(int) des(int) aes(partial) idea(int) blowfish(ptr) 
compiler: ccache gcc -fPIC -pthread -m64 -Wall -O3 -g -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DNDEBUG
                              op      op/s
 521 bits ecdh (nistp521)   0.0039s    259.5
```

**Comment by romen** [comment]
c4d2e27 adds a `CHANGES` entry.

Of course doing this will prevent clean cherry-pick to 1.1.1 and 1.1.0.

Once this PR gets approved I will open separate backporting PRs for 1.1.1 and 1.1.0.

**Comment by mspncp** [comment]
> [c4d2e27](https://github.com/openssl/openssl/commit/c4d2e2779bc40862670a436e8eac2aaecfe11256) adds a `CHANGES` entry.

> Of course doing this will prevent clean cherry-pick to 1.1.1 and 1.1.0.

Technically, the CHANGES file won't cherry-pick without merge conflict. But it's one which can trivially resolved by a human: Simply discard the incoming changes (`git checkout --ours CHANGES`) then edit the CHANGES file and re-add your entry at the correct location. Finally,  `git add CHANGES ; git cherry-pick --continue`.  So IMHO no need to open a separate pull request for it.

**Comment by romen** [comment]
> So IMHO no need to open a separate pull request for it. (_Edit:_ If it's the only merge conflict.)

Yes, I could trivially do that edit for the CHANGES file during the cherry-pick, but what is the policy about it? I tend to err on the over-conservative interpretation when in doubt.

**Comment by mspncp** [comment]
I've done it previously, and I know that @levitte does it sometimes, too. But AFAIK there is no official policy about that.

**Comment by romen** [comment]
Merged into master.

PR for 1.1.1 and 1.1.0 will show up soon as in both cases, in addition to conflicts on `CHANGES`, `test/build.info` requires manual intervention during cherry-pick.

Thanks all!

**Comment by mattcaswell** [comment]
> I've done it previously, and I know that @levitte does it sometimes, too. But AFAIK there is no official policy about that.

I will also manually resolve some merge conflicts as long as I think the fixes are trivial. Where the line is between trivial and non-trivial is probably a bit fuzzy. I would typically treat a "CHANGES" cherry-pick merge failure as trivial.

**Comment by levitte** [comment]
I'm of the same opinion as @mattcaswell on this.  But ultimately, I'd say that the comfort level of the one doing the merge, so if a backport PR was more suited to your comfort level, @romen, then that's perfectly fine.

**Pull request closed without merging**